oss-internship-2020/libtiff/example/main_sandboxed.cc

```diff
@@ -0,0 +1,277 @@
+                          int min_alpha, int max_alpha,
+                          const sapi::v::Array<unsigned> &buffer) {
+  // RGBA images are upside down - adjust for normal ordering
+  int adjusted_pixel = pixel % 128 + (127 - (pixel / 128)) * 128;
+  uint32 rgba = buffer[adjusted_pixel];
+
+  if (TIFFGetR(rgba) >= (uint32)min_red && TIFFGetR(rgba) <= (uint32)max_red &&
+      TIFFGetG(rgba) >= (uint32)min_green &&
+      TIFFGetG(rgba) <= (uint32)max_green &&
+      TIFFGetB(rgba) >= (uint32)min_blue &&
+      TIFFGetB(rgba) <= (uint32)max_blue &&
+      TIFFGetA(rgba) >= (uint32)min_alpha &&
+      TIFFGetA(rgba) <= (uint32)max_alpha) {
+    return 0;
+  }
+
+  std::cerr << "Pixel " << pixel << " did not match expected results.\n"
+            << "Got R=" << TIFFGetR(rgba) << " (expected " << min_red << ".."
+            << max_red << "), G=" << TIFFGetG(rgba) << " (expected "
+            << min_green << ".." << max_green << "), B=" << TIFFGetB(rgba)
+            << " (expected " << min_blue << ".." << max_blue
+            << "), A=" << TIFFGetA(rgba) << " (expected " << min_alpha << ".."
+            << max_alpha << ")\n";
+  return 1;
+}
+std::string GetFilePath(const std::string &dir, const std::string &filename) {
+  return sandbox2::file::JoinPath(dir, "test", "images", filename);
+}
+
+std::string GetCWD() {
+  char cwd[PATH_MAX];
+  getcwd(cwd, sizeof(cwd));
+  return cwd;
+}
+
+std::string GetFilePath(const std::string filename) {
+  std::string cwd = GetCWD();
+  auto find = cwd.rfind("build");
+
+  std::string project_path;
+  if (find == std::string::npos) {
+    std::cerr << "Something went wrong: CWD don't contain build dir. "
+              << "Please run tests from build dir or send project dir as a "
+              << "parameter: ./sandboxed /absolute/path/to/project/dir \n";
+    project_path = cwd;
+  } else {
+    project_path = cwd.substr(0, find);
+  }
+
+  return sandbox2::file::JoinPath(project_path, "test", "images", filename);
+}
+
+int main(int argc, char** argv) {
+  gflags::ParseCommandLineFlags(&argc, &argv, true);
+
+  std::string srcfile;
+  // "test/images/quad-tile.jpg.tiff"
+  std::string srcfilerel = "quad-tile.jpg.tiff";
+
+  if (argc < 2) {
+    srcfile = GetFilePath(srcfilerel);
+  } else {
+    srcfile = GetFilePath(argv[1], srcfilerel);
+  }
+
+  // without addDir to sandbox. to add dir use
+  // sandbox(absolute_path_to_dir, srcfile) or
+  // sandbox(absolute_path_to_dir). file and dir should be exists.
+  // srcfile must also be absolute_path_to_file
+  TiffSapiSandbox sandbox("", srcfile);
+
+  // initialize sapi vars after constructing TiffSapiSandbox
+  sapi::v::UShort h, v;
+  sapi::StatusOr<TIFF*> status_or_tif;
+  sapi::StatusOr<int> status_or_int;
+  sapi::StatusOr<tmsize_t> status_or_long;
+
+  auto status = sandbox.Init();
+  if (!status.ok()) {
+    fprintf(stderr, "Couldn't initialize Sandboxed API: %s\n", status);
+  }
```

Comment:
Error handling: Like in https://github.com/google/sandboxed-api/blob/master/oss-internship-2020/pffft/main_pffft_sandboxed.cc#L90, I suggest to refactor this to have a `LibTIFFMain()` function that returns an `absl::Status` and to use the `SAPI_RETURN_IF_ERROR` and `SAPI_ASSIGN_OR_RETURN` macros for error handling.
That way, you don't have to have `fprintf()`s everwhere can can handle errors in a single location.

Also, in case of an error this code does not return but continues, which is not what we want here.